src/Generating/EndGen.cpp

```diff
@@ -62,95 +59,118 @@ cEndGen::cEndGen(int a_Seed) :
+	for (int y = 0; y < DIM_Y; y++)
 	{
-		for (int x = 0; x < cChunkDef::Width; x++)
+		NOISE_DATATYPE ValY = static_cast<NOISE_DATATYPE>(2 * INTERPOL_Y * y - m_IslandThickness) / m_IslandThickness;
+		ValY = std::pow(ValY, 6);
+		for (int z = 0; z < DIM_Z; z++)
 		{
-			NOISE_DATATYPE noise = NoiseData[z * cChunkDef::Width + x];
+			for (int x = 0; x < DIM_X; x++)
+			{
+				NoiseData[x + DIM_X * z + DIM_X * DIM_Z * y] += ValY;
+			}  // for x
+		}  // for z
+	}  // for y
 
-			// The distance from spawn is used to create the void between the main island and the other islands.
-			double distanceFromSpawn = cChunkDef::RelativeToAbsolute({ x, 0, z }, a_ChunkCoords).Length();
+	// Upscale into real chunk size:
+	LinearUpscale3DArray(NoiseData, DIM_X, DIM_Z, DIM_Y, m_NoiseArray, INTERPOL_X, INTERPOL_Z, INTERPOL_Y);
+}
 
-			// The main island can get a different airthreshold. This way the other island can be more sparse while the main island
-			// is one big island.
-			if (distanceFromSpawn > m_MainIslandSize)
-			{
-				if (noise <= m_AirThresholdOtherIslands)
-				{
-					continue;
-				}
-				noise -= m_AirThresholdOtherIslands;
-			}
-			else
-			{
-				if (noise <= m_AirThresholdMainIsland)
-				{
-					continue;
-				}
-				noise -= m_AirThresholdMainIsland;
-			}
-			NOISE_DATATYPE voidOffset = VoidOffsetData[z * cChunkDef::Width + x];
 
 
-			double maxHeightLimit;
-			if (distanceFromSpawn > m_MainIslandSize * 3)
+
+
+void cEndGen::GenShape(cChunkCoords a_ChunkCoords, cChunkDesc::Shape & a_Shape)
+{
+	PrepareState(a_ChunkCoords);
+
+	int MaxY = std::min(static_cast<int>(1.75 * m_IslandThickness + m_IslandYOffset), cChunkDef::Height - 1);
+
+	// Choose which threshold to use depending on the distance from spawn.
+	double distanceFromSpawn = cChunkDef::RelativeToAbsolute({ cChunkDef::Width / 2, 0, cChunkDef::Width / 2 }, a_ChunkCoords).Length();
+	double minThreshold = distanceFromSpawn > m_MainIslandSize * 2 ? m_SmallIslandMinThreshold : m_MainIslandMinThreshold;
+	for (int z = 0; z < cChunkDef::Width; z++)
+	{
+		for (int x = 0; x < cChunkDef::Width; x++)
+		{
+			// Calculate the required treshold based on the distance from spawn.
+			// This way a void can be generated between the main island and the other islands.
+			double distanceFromSpawn = cChunkDef::RelativeToAbsolute({ x, 0, z }, a_ChunkCoords).Length();
+			double pow = std::pow((distanceFromSpawn - m_MainIslandSize) / m_MainIslandSize / 2, 3);
+			double mult = 3 * ((distanceFromSpawn - m_MainIslandSize) / m_MainIslandSize);
+			double threshold = std::min(pow - mult, minThreshold);
+
+			for (int y = 0; y < m_IslandYOffset; y++)
 			{
-				// The distance from spawn is so big we don't need to calculate the max height anymore.
-				// In fact, if we don't cut it off somewhere there is a chance the maxheight gets too big which
-				// can cause corrupted looking terrain.
-				maxHeightLimit = static_cast<float>(cChunkDef::Height);
+				a_Shape[y + x * 256 + z * 256 * 16] = 0;
```

Comment:
More magic numbers with available constants. I know that in this case, the code is longer but I think it might be helpful

Reply:
I think it might be useful to create a separate function in `cChunkDesc` that sets and gets the shape values. Now all generators manually calculate the index.

Reply:
Or in the common grandparent object

Reply:
`Shape` is a typedef of an array in `cChunkDesc`, so I think it makes sense to put the functions next to that.